nordic: nRF54L: Add the nRF54L05 and nRF54L10 ICs

carlescufi opened the pull request:
These two new ICs are variants of the nRF54L15 with different memory
    sizes:
    
    - nRF54L05: 500KB RRAM, 96KB RAM
    - nRF54L10: 1022KB RRAM, 192KB RAM
    - nRF54L15: 1524KB RRAM, 256KB RAM

carlescufi committed:
soc: nordic: nRF54L: Consolidate common Kconfig options

There are many common options to all ICs of the 54L series. Consolidate
them in a single entry so that they do not need to be re-typed for each
SoC series member.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
soc: nordic: Introduce the nRF54L05 and nRF54L10

These two new ICs are variants of the nRF54L15 with different memory
sizes:

- nRF54L05: 500KB RRAM, 96KB RAM
- nRF54L10: 1022KB RRAM, 192KB RAM
- nRF54L15: 1524KB RRAM, 256KB RAM

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
boards: nordic: nrf54l15dk: Rename the board common file

Use the rather logical convention for the name that is applied to other
Nordic boards: <board>_common.dtsi for definitions that are common to
the board itself (LEDs, buttons, etc).

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
boards: nordic: nRF54L15DK: Add basic support for the L05 and L10 ICs

The nRF54L05 and nRF54L10 are identical to the nRF54L15 except for their
memory sizes. Add support for emulating those ICs on the nRF54L15DK.
This commit only adds support for the main application core. Support for
the FLPR core may be added later.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented on boards/nordic/nrf54l15dk/Kconfig.nrf54l15dk:
```diff
@@ -2,6 +2,8 @@
+	select SOC_NRF54L05_CPUAPP if BOARD_NRF54L15DK_NRF54L05_CPUAPP
+	select SOC_NRF54L10_CPUAPP if BOARD_NRF54L15DK_NRF54L10_CPUAPP
 	select SOC_NRF54L15_CPUAPP if BOARD_NRF54L15DK_NRF54L15_CPUAPP
 	select SOC_NRF54L15_CPUFLPR if BOARD_NRF54L15DK_NRF54L15_CPUFLPR || \
```
Yes, it's in the commit message. I prefer to follow-up with a PR for that, but if you'd like I can add them here too.

carlescufi replied:
Discussed internally and we will add them later

carlescufi commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -1056,10 +1056,12 @@
-    #include <nrfx_config_nrf54l15_flpr.h>
+#elif (defined(NRF54L05_XXAA) || defined(NRF54L10_XXAA) || defined(NRF54L15_XXAA)) \
+	&& defined(NRF_APPLICATION)
+    #include <nrfx_config_nrf54l_05_10_15_application.h>
```
It's not autogenerated, is it? @anangl?

gmarull replied:
If they are zero-diff, I'd consolidate them in a single file (templates are in HAL repo)

masz-nordic replied:
They should be copied from https://github.com/zephyrproject-rtos/hal_nordic/tree/master/nrfx/templates which are auto-generated.

anangl replied:
These are anyway just templates, so it's not mandatory to use them exactly as they are provided.
In may opinion, it makes perfect sense to use one file instead of three if their actual content is the same. Wouldn't it be possible to somehow tweak this auto-generation (add an extra step or something) to end up with such common file as a template in nrfx?

masz-nordic replied:
>These are anyway just templates, so it's not mandatory to use them exactly as they are provided.

Correct, but who is then responsible for keeping files generated and ones manually created here in sync? We've seen cases where incorrect/outdated config is causing issues.

I agree tweaking auto-generation is an option, but the gain is minimal. So for now I'd say we should use files provided.

Also, usage of nrfx_config in Zephyr seems to be overly complicated, I will propose a PR in the future to rework this.

carlescufi commented on soc/nordic/nrf54l/Kconfig:
```diff
@@ -20,20 +21,35 @@ config SOC_NRF54L15_CPUAPP
-	select HAS_POWEROFF
+config SOC_NRF54L_05_10_15_CPUAPP
+	bool
+	select SOC_NRF54LX_CPUAPP_COMMON
```
I thought about that, but I liked being future-proof. Let's hear from more reviewers before I make the change.

carlescufi replied:
@gmarull @bjarki-andreasen @anangl thoughts on this?

gmarull commented on dts/arm/nordic/nrf54l05_cpuapp.dtsi:
```diff
@@ -0,0 +1,8 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <nordic/nrf54l05.dtsi>
+#include <nordic/nrf54l_05_10_15_cpuapp.dtsi>
```
includes should go from less to more specific?

carlescufi replied:
The top-level include is the generic SoC, then some shared definitions specific to the cpuapp core. I think that's actually correct?

gmarull commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l05_cpuapp_defconfig:
```diff
@@ -0,0 +1,29 @@
+# Enable hardware stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# MPU-based null-pointer dereferencing detection cannot
+# be applied as the (0x0 - 0x400) is unmapped for this target.
+CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
+
+# Enable Cache
+CONFIG_CACHE_MANAGEMENT=y
+CONFIG_EXTERNAL_CACHE=y
+
+# Start SYSCOUNTER on driver init
+CONFIG_NRF_GRTC_START_SYSCOUNTER=y
```
for future: we should likely move these to soc defconfig, so that custom boards stay safe (and simpler) if this is not set.

carlescufi replied:
Agreed! Made a note, will open a PR soon.

miha-nordic commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l05_cpuapp.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15dk/nrf54l15/cpuapp
```
Shouldn't be "nrf54l15dk/nrf54l05/cpuapp"?
And name change below.

miha-nordic replied:
Shouldn't be "nrf54l15dk/nrf54l05/cpuapp"?
And name change below?

miha-nordic commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l10_cpuapp.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15dk/nrf54l15/cpuapp
```
Shouldn't be "nrf54l15dk/nrf54l10/cpuapp"?
And name change below.

miha-nordic commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l05_cpuapp.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15dk/nrf54l15/cpuapp
```
Shouldn't be "nrf54l15dk/nrf54l05/cpuapp"?
And name change below?

carlescufi replied:
fixed, thanks!

miha-nordic commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l10_cpuapp.yaml:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15dk/nrf54l15/cpuapp
```
Shouldn't be "nrf54l15dk/nrf54l10/cpuapp"?
And name change below?

carlescufi replied:
fixed, thanks!

carlescufi commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -55,6 +55,12 @@ zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54H20_ENGB_CPUFLPR NRF54H20_ENGB_
 zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L15            NRF54L15_XXAA)
 zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L15_CPUAPP     NRF_APPLICATION)
 zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L15_CPUFLPR    NRF_FLPR)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L10            NRF54L10_XXAA)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L10_CPUAPP     NRF_APPLICATION)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L10_CPUFLPR    NRF_FLPR)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L05            NRF54L05_XXAA)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L05_CPUAPP     NRF_APPLICATION)
+zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54L05_CPUFLPR    NRF_FLPR)
```
Actually the `.svd` were already correctly sorted

anangl commented on dts/common/nordic/nrf54l05.dtsi:
```diff
@@ -0,0 +1,27 @@
+};
+
+&cpuflpr_sram {
+	reg = <0x2002f000 DT_SIZE_K(24)>;
```
This is incorrect. The start address should be 0x20012000, right after `cpuapp_sram` (0x2002f000 is out of range of the available SRAM).
I think the memory nodes, at least the FLPR ones, should not be defined in the common file but rather in the SoC specific files (because the start addresses are part of their names).

anangl commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l05_cpuapp.yaml:
```diff
@@ -0,0 +1,24 @@
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: nrf54l15dk/nrf54l05/cpuapp
+name: nRF54l15-DK-nRF54l05-Application
```
```suggestion
name: nRF54L15-DK-nRF54L05-Application
```
Applies also to L10 and L15.

anangl commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l05_cpuapp.dts:
```diff
@@ -0,0 +1,52 @@
+};
+
+&cpuapp_rram {
+	partitions {
```
These partitions sum up to 500 kB, but `cpuapp_rram` size is 470 kB. Similar problem for nrf54l10.

carlescufi replied:
But FLPR is not supported yet on this board. What do you suggest?

aescolar committed:
manifest: Update nRF hw models to latest

Update the HW models module to:
42737c8ec8485987c7c9b0262b136de623e1ded2

Including the following:
42737c8 TIMER: Support devices without TASK_SHUTDOWN nrfx3.9 MDK 8.68
5fe6873 54 UARTE: Add frametimeout functionality
dc086d7 UARTE: Add basic 54 support
b046745 UARTE: Support better not having UART functionality
1c5f58c README: Mention the nRF54L15 models cover the L10 and L05
597c7d0 TEMP: Also build hal replacement for 54 and define NRF_TEMP_NS/S
fb2ca83 Makefile: Let's build libraries (specially HAL) as pic
200a1e3 Makefiles: move some common options to common snippet

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

nordic-piks commented on soc/nordic/soc.yml:
```diff
@@ -21,6 +21,14 @@ family:
+          - name: nrf54l10
+            cpuclusters:
+              - name: cpuapp
+              - name: cpuflpr
```
Down in this file there should be added elements for nrf54l05 and nrf54l10, the same as there is nrf54l15
(so that reset, erase and recovery will work correctly)

anangl commented on soc/nordic/common/vpr/Kconfig.sysbuild:
```diff
@@ -4,7 +4,7 @@
 	bool "VPR launcher"
 	default y
-	depends on (SOC_NRF54H20_CPUPPR || SOC_NRF54H20_CPUFLPR || SOC_NRF54L15_CPUFLPR || SOC_NRF9280_CPUPPR)
+	depends on (SOC_NRF54H20_CPUPPR ||  SOC_NRF54H20_CPUFLPR ||  SOC_NRF54L_05_10_15_CPUFLPR || SOC_NRF9280_CPUPPR)
```
Actually, this could be simplified to:
```suggestion
	depends on RISCV_CORE_NORDIC_VPR
```

anangl replied:
This change should be reverted, since an option like `SOC_NRF54L_05_10_15_CPUFLPR` does not exist. But since FPLRs are not supported for L05 and L10 boards, nothing needs to be changed here.

anangl commented on drivers/misc/nordic_vpr_launcher/nordic_vpr_launcher.c:
```diff
@@ -14,7 +14,7 @@
 
 #include <hal/nrf_vpr.h>
-#if defined(CONFIG_SOC_NRF54L15_CPUAPP) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
+#if defined(CONFIG_SOC_NRF54L_05_10_15_CPUAPP) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```
```suggestion
#if defined(CONFIG_SOC_NRF54L_CPUAPP_COMMON) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```

anangl commented on drivers/misc/nordic_vpr_launcher/nordic_vpr_launcher.c:
```diff
@@ -41,7 +41,7 @@ static int nordic_vpr_launcher_init(const struct device *dev)
 #endif
 
-#if defined(CONFIG_SOC_NRF54L15_CPUAPP) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
+#if defined(CONFIG_SOC_NRF54L_05_10_15_CPUAPP) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```
```suggestion
#if defined(CONFIG_SOC_NRF54L_CPUAPP_COMMON) && !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE)
```

anangl commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -56,7 +56,7 @@
 #define IRQ0_PRIO	IRQ_DEFAULT_PRIORITY
 #define IRQ1_PRIO	0x0
-#elif defined(CONFIG_SOC_NRF54L15_CPUFLPR)
+#elif defined(CONFIG_SOC_NRF54L_05_10_15_CPUFLPR)
```
```suggestion
#elif defined(CONFIG_SOC_SERIES_NRF54LX) && defined(CONFIG_RISCV_CORE_NORDIC_VPR)
```

anangl commented on soc/nordic/nrf54l/Kconfig:
```diff
@@ -20,19 +21,29 @@ config SOC_NRF54L15_CPUAPP
+
+config SOC_NRF54L05_CPUFLPR
+	depends on RISCV_CORE_NORDIC_VPR
+	select SOC_NRF54L_05_10_15_CPUFLPR
```
```suggestion
```

anangl commented on soc/nordic/nrf54l/Kconfig:
```diff
@@ -20,19 +21,29 @@ config SOC_NRF54L15_CPUAPP
+
+config SOC_NRF54L10_CPUFLPR
+	depends on RISCV_CORE_NORDIC_VPR
+	select SOC_NRF54L_05_10_15_CPUFLPR
```
```suggestion
```

anangl commented on soc/nordic/nrf54l/Kconfig:
```diff
@@ -20,19 +21,29 @@ config SOC_NRF54L15_CPUAPP
 
 config SOC_NRF54L15_CPUFLPR
 	depends on RISCV_CORE_NORDIC_VPR
+	select SOC_NRF54L_05_10_15_CPUFLPR
```
```suggestion
```

anangl commented on soc/nordic/nrf54l/Kconfig:
```diff
@@ -9,7 +9,7 @@ config SOC_SERIES_NRF54LX
 	select HAS_SEGGER_RTT if ZEPHYR_SEGGER_MODULE
 
-config SOC_NRF54LX_CPUAPP_COMMON
+config SOC_NRF54L_CPUAPP_COMMON
```
This correction should be done in the first commit.

nordic-piks reviewed (approved):
soc/nordic/soc.yml LGTM.

nika-nordic commented:
needs rebase as prerequisite nrfx 3.8.0 was just merged: https://github.com/zephyrproject-rtos/zephyr/pull/79857

Also, hal_nordic PR with nrfx 3.9.0 needs to be added here: https://github.com/zephyrproject-rtos/hal_nordic/pull/257 so that twister can become green

carlescufi commented:
last push: just a rebase to resolve conflicts

carlescufi committed:
manifest: hal_nordic: Update to nrfx 3.9.0

See https://github.com/zephyrproject-rtos/hal_nordic/pull/257.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
> needs rebase as prerequisite nrfx 3.8.0 was just merged: #79857
> 
> Also, hal_nordic PR with nrfx 3.9.0 needs to be added here: [zephyrproject-rtos/hal_nordic#257](https://github.com/zephyrproject-rtos/hal_nordic/pull/257) so that twister can become green

All done, thanks!

nika-nordic commented:
@aescolar 

Looks like babblesim alignment will be needed:
> /__w/zephyr/zephyr/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:1727:45: error: ‘NRF_TIMER_TASK_SHUTDOWN’ undeclared (first use in this function); did you mean ‘NRF_TIMER_HAS_SHUTDOWN’?
 1727 |         nrf_timer_task_trigger(EVENT_TIMER, NRF_TIMER_TASK_SHUTDOWN);
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~
      |                                             NRF_TIMER_HAS_SHUTDOWN

TASK_SHUTDOWN is not available on all devices now. This is how it is handled in nrfx_timer driver:

```
void nrfx_timer_disable(nrfx_timer_t const * p_instance)
{
    NRFX_ASSERT(m_cb[p_instance->instance_id].state != NRFX_DRV_STATE_UNINITIALIZED);

#if NRF_TIMER_HAS_SHUTDOWN
    nrfy_timer_task_trigger(p_instance->p_reg, NRF_TIMER_TASK_SHUTDOWN);
#else
    nrfy_timer_task_trigger(p_instance->p_reg, NRF_TIMER_TASK_STOP);
    nrfy_timer_task_trigger(p_instance->p_reg, NRF_TIMER_TASK_CLEAR);
#endif
    m_cb[p_instance->instance_id].state = NRFX_DRV_STATE_INITIALIZED;
    NRFX_LOG_INFO("Disabled instance: %d.", p_instance->instance_id);
}
```

aescolar committed:
Bluetooth: Controller: nRF: Support MDK 8.68

TASK_SHUTDOWN was deprecated in newer SOCs and now removed
in MDK 8.68 (nrfx 3.9.0)

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

masz-nordic reviewed (dismissed):
-1 for manually created nrfx_config

aescolar commented:
> Looks like babblesim alignment will be needed:
> [..]
> Looks like other parts of Zephyr needs alignment as well:

This PR updates both the HW models and the BT controller HAL so it works with either this or the old nrfx:
https://github.com/zephyrproject-rtos/zephyr/pull/81644
You are welcome to pick those 2 commits into this PR. But I would get that other PR merged ahead of this if this one drags to minimize cross dependencies.
Note there is runtime failures with this latest HAL both in the controller (for 54L15) and the UART driver (CC @cvinayak  @nordic-krch )

nika-nordic commented:
> Note there is runtime failures with this latest HAL both in the controller (for 54L15) and the UART driver

If any of these utilize nrfx_clock.c or nrfx_power.c driver the culprit may be changed CLOCK and POWER IRQ number. It used to be IRQn==270 but now it is IRQn==261. nrfx_clock.c and nrfx_power.c enables IRQn==261 now, but in Zephyr dts `270` is still used, so this should be aligned

aescolar committed:
manifest: Update nRF hw models to version compatible with nrfx3.9

Update the HW models module to:
3cfca0192ff84da919e9bc7978bcc2239cd6a395

This includes 2 changes:
* 3cfca01 UART: Model more accurately TASKS_FLUSH behaviour
* 25cbd28: 54L15: Change CLOKPOWER IRQ line (MDK 8.68, nrfx 3.9.0)

    WARNING! This change breaks backwards compatibility with SW
    which expected this line to be 270 for 54L15 devices,
    and therefore with MDKs < 8.68 & nrfx < 3.9.0.
    This change requires updating to the MDK >=8.68 and nrfx >=3.9.0
    if building using the 54L15 models.
    52 and 53 models are not affected.

    Change the CLOCK_POWER interrupt line from 270 to 261.
    This has changed in the latest MDK (8.68) and the new
    value is used in the latest nrfx (3.9.0) drivers.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
